### ISSUE TITLE
Optionally optimize map size by discarding unmodified mapblocks

### ISSUE THREAD

**Issue by archfan7411** [minetest/minetest#13604]
## Problem

Minetest maps can easily grow into the tens of gigabytes of size. For server owners hosting on VPSes, this can rapidly balloon the costs of hosting.

In many cases, players explore far abroad without modifying the map at all. To the storage-conscious host, storing these unmodified mapblocks is inefficient at best.

## Solutions

Provide a configuration option, such as `discard_unmodified_blocks`, which would result in the server not writing any mapblock to the DB unless it's been changed. **I am willing to try implementing this myself**, but first want to get an idea of what people think.

## Potential issues

Discarding unmodified blocks would result in undesirable behavior for anyone who isn't storage-limited, as it would increase CPU usage overall when mapblocks are regenerated. There may also be (potentially conclusive) issues with non-deterministic mapgens, decorations, or Lua mapgens, depending on how this is implemented.

However, since this would be disabled by default, I don't think these pose great problems.

## Alternatives

Do nothing, I guess.

## Additional context

`debug.txt`, which can also balloon into the gigabytes, already has options in `minetest.conf` for splitting it and otherwise managing it over time.

**Comment by SmallJoker** [comment]
Where would you draw the line where a mapblock is modified or not? ABMs, LBMs and node timers may automatically modify mapblocks without any player interaction. The perhaps simplest way is to mark the mapblock for saving as soon as a player dug a node or killed an entity within (which might be spawned automatically on generation).

By the way: I have the feeling I already saw this request in another issue but cannot find any as of now.

**Comment by archfan7411** [comment]
> Where would you draw the line where a mapblock is modified or not? ABMs, LBMs and node timers may automatically modify mapblocks without any player interaction.

Good point; I was already thinking that LBMs would be ignored. For *my* usecases, I would probably want to ignore ABMs and node timers too, but if a node/entity is generated, starts doing something, and then suddenly gets erased from existence, that could cause some very confusing issues. Looks like this may not be as easy as I was hoping.

> By the way: I have the feeling I already saw this request in another issue but cannot find any as of now.

I seem to recall reading one in the past as well, and I think it also explored the option of marking mapblocks.

**Comment by BuckarooBanzay** [comment]
somewhat related: this is already possible (with a slightly different approach) using an external application: https://github.com/minetest-go/mapcleaner

Some numbers for the last time i ran this on the pandorabox server:
* Before: 70GB / 127M Mapblocks
* After: 18GB / 26M Mapblocks

**Comment by archfan7411** [comment]
I appreciate the link, but this doesn't accomplish the same thing (it wouldn't be transparent to the end user, as it would remove terraforming / noob dirt houses etc.)

**Comment by Desour** [comment]
I remember some similar discussion on the forums, but I can't find the topic right now.

What I suggested there:
If mapgen is deterministic, one could do mapgen again for old mapchunks, and only store the diffs. On re-mapgen of these chunks, one would then after mapgen apply this diff.
But that's a bit too complicated for now.

**Comment by sfan5** [comment]
This (both what OP is suggesting and what DS said) is harder than you think.

While the base mapgen is deterministic, stuff above it is often not so much. No guarantees for what mods do of course.
Also you can only generate the map in chunks larger than one mapblock.
Plus when one of these chunks generate they can modify the directly neighboring mapblocks.

If you've dealt with all that the next question is tracking modifications.
As SmallJoker already said, do you track LBMs and ABMs?
If an entity makes its way into the block, do you not delete it?
That would be bad if the mapgen process spawns entities so let's say you wanted to exclude those. How would that be reliably tracked?
Do not assume that you can get by with only counting explicit node digging/placement either. There are plenty of ways to affect a mapblock without doing that.

**Comment by archfan7411** [comment]
I think I would expect this feature to discard ABMs/LBMs and any entities. I'm aware that entities are tracked separately from the mapblock, and so that would increase complexity. 

Tracking what modifications are made by ABMs/LBMs/mapgen and which aren't sounds like it would require additional logic on top of the aforementioned, where ABMs/LBMs/etc would specifically not "mark" the mapblock as changed.

This feature innately isn't applicable everywhere, so I would argue that it not supporting some edge cases is reasonable.

**Comment by rubenwardy** [comment]
I suggest having a flag on mapblocks to track this along with last modification date, and then let server owners manually prune the map database

**Comment by archfan7411** [comment]
I'd argue that having a setting accomplishes much the same thing in terms of owner choice; but I'd certainly be happy with just marked mapblocks too.

**Comment by lhofhansl** [comment]
Somewhat related... It has been irking me for a while that we have to save the entire block data each time the block's timestamp changes, instead of having a slightly different schema which stores the timestamp in a separate column.

**Comment by appgurueu** [comment]
I imagine that many blocks are close enough to be worth generating for larger viewing ranges, but too far away for *anything* to happen to them (if the game developer is careful). The default ABM range is just 4 blocks for example (64 nodes), whereas normal view ranges are twice or quadruple that.

Let $r$ be the radius around the player where we emerge blocks. The number of blocks generated will scale roughly quadratically in $r$, linearly in the length of the path. When we go from "emerge range" to "ABM range" when it comes to keeping blocks, I would thus estimate a fraction on the order of $\frac{1}{4}$ to $\frac{1}{16}$ of the blocks. That'd already be a big improvement with a very simple strategy.

> But anyway, storing this information is a good start.

Agreed on this. The engine should know when a block has not been touched at all since generation, which a careful game dev can guarantee [^lbm-timer], moreso if the game is particularly simple. A timestamp is useful because it lets you run (possibly much more efficient) catch-up logic if necessary.

A more general solution might be some flag to inform the engine that a map modification need not be kept, but this is more difficult to design than a simple "when was this block last touched" / "was this block last touched since generation".

[^lbm-timer]: LBMs and node timers can of course be used to dirty all mapblocks immediately after generation, but that's on the game dev. Ideally a generated block should be ready as-is; LBMs are mostly for migrations from older versions. Node timers should be confined to a range similar to the ABM range.

> You could have flags for whether the player has done any dig/place/punch.

This kind of thing should be possible with a mod and some form of mapblock metadata #16125. To make it convenient, we should probably expose a single Lua API function that catches all player interactions (`TOSERVER_INTERACT`). That is also a good idea for other reasons, for example to have one central point where an anticheat mod can inspect suspicious interactions.
